### PR TITLE
find-tarballs.nix: Include patches

### LINKED ISSUE #186831: find-tarballs.nix not including patches
### Describe the bug

The script https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/find-tarballs.nix should report all sources used to build a package. It currently lacks support for patches?

If we run find-tarballs on nethack, we should find the patch defined in nethack https://github.com/NixOS/nixpkgs/blob/master/pkgs/games/nethack/default.nix#L27-L34 named `81d73ce417dda6a98e2e918e06922e68b67c53f7.patch`

### Steps To Reproduce

```
nix-instantiate --json --eval --strict "<nixpkgs/maintainers/scripts/find-tarballs.nix>" --arg expr "(import <nixpkgs>{}).nethack" | nix-shell -p jq --run jq | grep "81d73ce417dda6a98e2e918e06922e68b67c53f7.patch"```

### Expected behavior

we should get the url to download `81d73ce417dda6a98e2e918e06922e68b67c53f7.patch` in the output, along with its name, hash and type.

### Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 5.18.11-xanmod1, NixOS, 22.11 (Raccoon), 22.11.20220808.39d7f92`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.10.3`
 - channels(root): `""`
 - nixpkgs: `/nix/store/8dvbcv90f940fg9biwz42xbh5d3xpjav-source`
```

### PR DESCRIPTION
**Pull request by rapenne-s**
###### Description of changes

I'd like to be able to download all sources files required to build a package while being offline, but also check upstream is reachable.

We can currently have `find-tarballs.nix` to obtain a list of urls / hash / name for all distfiles of a given packages, but it's skipping patches files defined with `fetchpatch`. This is so because they have a "postFetch" attribute containing a script that must be run on the downloaded file to sanitize it, the file hash corresponds to the sanitized output.

I'm proposing to include the patches files to the list and to add an attribute `isPatch` to the result which can be used to differentiate between patches and non-patches. The other commit in the PR include changes to find-tarballs.nix consumer copy-tarballs.nix by adding the code to skip patches files because this would require some work to support them.

This fixes https://github.com/NixOS/nixpkgs/issues/186831

With this change, it's now possible to download all sources for a package by iterating over the json output of find-tarballs, and using nix-instantiate to either call fetchurl or fetchpatch.

Here is a sample of Perl code I'm using
```perl
    if ($val->{isPatch}) {
        $return = `nix-instantiate '<nixpkgs>' -A fetchpatch --argstr url "$val->{url}" --argstr name "$val->{name}" --argstr sha256 "$val->{hash}"`;
        print $return;
    } else {
        $return = `nix-instantiate '<nixpkgs>' -A fetchurl --argstr url "$val->{url}" --argstr sha256 "$val->{hash}"`;
        print $return;
    }
```

### CODE REVIEW

**Comment by rapenne-s** [comment]
I wrote the tool https://tildegit.org/solene/nixpkgs-mirror-tarballs to easily download all sources required to build a list of packages.

**Comment by roberth** [comment]
@edolstra @shlevy Does this script have any tests or documentation? Why don't we copy _all_ fixed outputs of Nixpkgs?

**Comment by rapenne-s** [comment]
From a `git grep` both script doesn't seem to be documented or be tested.

I overlooked `find-tarballs.nix` is used in `pkgs/top-level/make-tarball.nix` to generate a json with all tarballs to download for a release, it seems to be used to feed Hydra. Maybe hydra is using `copy-tarballs.nix` with this file, so the changes would have no impact.

**Comment by LunNova** [comment]
Is not applying postFetch a problem?

**Comment by rapenne-s** [comment]
> Is not applying postFetch a problem?

if you don't apply postFetch, the downloaded file will have a different hash than expected and won't be usable

**Comment by AndersonTorres** [comment]
> if you don't apply postFetch, the downloaded file will have a different hash than expected and won't be usable

In the sense of "bit-per-bit" software preservation, is the downloaded file kept? Or is it modified in place?

File: maintainers/scripts/copy-tarballs.pl [hunk 1]
```diff
@@ -189,6 +190,11 @@ sub uploadFile {
             next;
         }
 
+        if ($isPatch) {
+            print STDERR "skipping $url (support for patches is missing)\n";
```
Comment (AndersonTorres) [inline maintainers/scripts/copy-tarballs.pl:194]: Basically a TODO?

**Review by AndersonTorres — COMMENTED** [review]
I am seriously interested on this!

File: maintainers/scripts/copy-tarballs.pl (hunk 1 above)
Comment (rapenne-s) [reply maintainers/scripts/copy-tarballs.pl:194]: yes it's a TODO because it would be nice to have a copy on hydra I suppose

**Comment by rapenne-s** [comment]
> > if you don't apply postFetch, the downloaded file will have a different hash than expected and won't be usable
> 
> In the sense of "bit-per-bit" software preservation, is the downloaded file kept? Or is it modified in place?

no, the downloaded file is not kept, we only keep the result file and its checksum, we don't know the checksum of the original file.

If you are interested, it's not that hard to download the raw patch file, but it's not possible to verify its integrity except if the `postFetch` applied on it give the expected hash.

**Pull request merged**